### ISSUE TITLE
The polling thread is not awaken by a pipe write

### ISSUE THREAD

**Issue by jxyang** [deislabs/mystikos#318]
This is discovered while debugging memcached. It creates 4 worker threads. Each thread creates a pipe and is blocked on `poll(NULL, 0, -1)` on the input end of the pipe after creation. When an incoming connection is established, the master thread assigns the socket to the first idle thread, and writes to the output end of the thread's pipe to wake it up.

Mystikos tries to wake up a thread that is blocked on `poll(NULL, 0, -1)` on the host side with a special pipe  `_wakefds`. But this pipe is shared by all threads. So the attempt above in memcached actually resulted in the wrong thread getting awaken.

At the end, the thread which owns the socket is still sleeping, and the thread that doesn't know the socket is awaken. The final result is we can connect to the memcached server, but any command sent to it from the client gets no response.

**Comment by jxyang** [comment]
This is a small test case to repro (derived from tests/pollpile):
```c
#include <assert.h>
#include <limits.h>
#include <poll.h>
#include <pthread.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define ITERATIONS 1

#ifdef TRACE
#define T(EXPR) EXPR
#else
#define T(EXPR)
#endif

static const char alphabet[] = "abcdefghijklmnopqrstuvwxyz";

static int pipefd[2][2];

static void _sleep_msec(uint32_t msec)
{
    struct timespec ts;
    ts.tv_sec = (uint64_t)msec / 1000;
    ts.tv_nsec = ((int64_t)msec % 1000) * 1000000;
    nanosleep(&ts, NULL);
}

static void* _reader(void* arg)
{
    size_t thread_id = (size_t)arg;

    for (size_t i = 0; i < ITERATIONS; i++)
    {
        struct pollfd fds;
        int n;
        char buf[sizeof(alphabet)];

        fds.fd = pipefd[thread_id][0];
        fds.events = POLLIN;

        n = poll(&fds, 1, -1);

        printf("Thread %ld after poll\n", thread_id);
        assert(n == 1);

        printf("Thread %ld reading:  %zu\n", thread_id, i);
        ssize_t count = read(pipefd[thread_id][0], buf, sizeof(buf));

        if (count != sizeof(alphabet))
            printf("count=%zu\n", count);
        assert(count == sizeof(alphabet));
        assert(strcmp(buf, alphabet) == 0);
    }

    return NULL;
}

static void* _writer(void* arg)
{
    size_t thread_id = (size_t)arg;

    for (size_t i = 0; i < ITERATIONS; i++)
    {
        printf("write to thread %ld: %zu\n", thread_id, i);
        ssize_t n = write(pipefd[thread_id][1], alphabet, sizeof(alphabet));

        if (n != sizeof(alphabet))
            printf("n=%zu\n", n);

        assert(n == sizeof(alphabet));
    }

    return NULL;
}

int main(int argc, const char* argv[])
{
    pthread_t reader[2];
    pthread_t writer;

    assert(pipe(pipefd[0]) == 0);
    assert(pipe(pipefd[1]) == 0);

    assert(pthread_create(&reader[1], NULL, _reader, (void*)0) == 0);
    assert(pthread_create(&reader[2], NULL, _reader, (void*)1) == 0);
    _sleep_msec(10000);
    assert(pthread_create(&writer, NULL, _writer, (void*)0) == 0);

    assert(pthread_join(writer, NULL) == 0);
    assert(pthread_join(reader[0], NULL) == 0);
    printf("===  after join reader 0\n");

    assert(close(pipefd[0][0]) == 0);
    assert(close(pipefd[0][1]) == 0);

    printf("=== passed test (%s)\n", argv[0]);

    return 0;
}
```

Writing to the pipe belongs to thread 0 should wake up thread 0, but the following output indicates thread 1 was awaken:

```
write to thread 0: 0
Thread 1 after poll
Assertion failed: n == 1 (pollpipe.c: _reader: 50)
2021-04-26T20:58:23+0000.040402Z [(H)ERROR] tid(0x7f7f409bfb80) | :OE_ENCLAVE_ABORTING [/home/xuejun/mikemyst/third_party/openenclave/openenclave/host/calls.c:_call_enclave_function_impl:56]
/home/xuejun/mikemyst/build/bin/myst: error: failed to enter enclave: result=OE_ENCLAVE_ABORTING
Makefile:22: recipe for target 'tests' failed
```

**Comment by jxyang** [comment]
Fixed by #324
